shell/platform/darwin/macos/framework/Source/FlutterTextInputSemanticsObject.mm

```diff
@@ -0,0 +1,131 @@
+  _node->GetDelegate()->AccessibilityPerformAction(data);
+}
+
+- (BOOL)becomeFirstResponder {
+  if (_plugin.client == self && [_plugin isFirstResponder]) {
+    // This text field is already the first responder.
+    return YES;
+  }
+  BOOL result = [super becomeFirstResponder];
+  if (result) {
+    _plugin.client = self;
+    // The default implementation of the becomeFirstResponder will change the
+    // text editing state. We need to manually set it back.
+    NSString* textValue = @(_node->GetStringAttribute(ax::mojom::StringAttribute::kValue).data());
+    int start = _node->GetIntAttribute(ax::mojom::IntAttribute::kTextSelStart);
+    int end = _node->GetIntAttribute(ax::mojom::IntAttribute::kTextSelEnd);
+    NSRange selection;
+    if (start > 0 && end > 0) {
+      selection = NSMakeRange(start, end - start);
+    } else {
+      selection = NSMakeRange([self stringValue].length, 0);
+    }
+    [self updateString:textValue withSelection:selection];
+  }
+  return result;
+}
+
+@end
+
+namespace flutter {
+
+FlutterTextPlatformNode::FlutterTextPlatformNode(FlutterPlatformNodeDelegate* delegate,
+                                                 FlutterEngine* engine) {
+  Init(delegate);
+  engine_ = engine;
+  native_text_field_ =
+      [[FlutterTextField alloc] initWithPlatformNode:this
+                                     withFieldEditor:engine.viewController.textInputPlugin];
+  native_text_field_.bezeled = NO;
+  native_text_field_.drawsBackground = NO;
+  native_text_field_.bordered = NO;
+  native_text_field_.focusRingType = NSFocusRingTypeNone;
+  [engine.viewController.view addSubview:native_text_field_
+                              positioned:NSWindowBelow
+                              relativeTo:engine.viewController.flutterView];
+}
+
+FlutterTextPlatformNode::~FlutterTextPlatformNode() {
+  if ([native_text_field_ isDescendantOf:engine_.viewController.view]) {
+    [native_text_field_ removeFromSuperview];
+  }
+  native_text_field_ = nil;
+  engine_ = nil;
+}
+
+gfx::NativeViewAccessible FlutterTextPlatformNode::GetNativeViewAccessible() {
+  return native_text_field_;
+}
+
+NSRect FlutterTextPlatformNode::GetWindowFrame() {
+  FlutterPlatformNodeDelegate* delegate = (FlutterPlatformNodeDelegate*)GetDelegate();
+  bool offscreen;
+  auto bridge_ptr = delegate->GetOwnerBridge().lock();
+  gfx::RectF bounds = bridge_ptr->RelativeToGlobalBounds(delegate->GetAXNode(), offscreen, true);
+
+  // Converts to NSRect to use NSView rect conversion.
+  NSRect ns_local_bounds = NSMakeRect(bounds.x(), bounds.y(), bounds.width(), bounds.height());
+  // The macOS XY coordinates start at bottom-left and increase toward top-right,
+  // which is different from the Flutter's XY coordinates that start at top-left
+  // increasing to bottom-right. Therefore, We need to flip the y coordinate when
+  // we convert from flutter coordinates to macOS coordinates.
```

Comment:
Change this whole sentence to "Flip the y coordinate to convert from Flutter coordinates to macOS coordinates." (See go/avoidwe)